### PR TITLE
Moves keyring stuffs to a package to save data locally

### LINKED ISSUE #299: Create a generic package to manage the client keyring
# New feature request

## Describe the proposal

As the CLI is growing in size each time more we are saving small pieces of data on the client machine using the keyring library, like the token, url and [soon the status of the CLI](https://github.com/calyptia/cli/issues/229).

as example:
https://github.com/calyptia/cli/blob/f69d77ada93d977230fbe1737d33de77e0ea20a5/cmd/calyptia/auth.go#L23-L49
https://github.com/calyptia/cli/blob/f69d77ada93d977230fbe1737d33de77e0ea20a5/cmd/calyptia/config_url.go#L70-L96

We already have a duplicated code to the URL part and i caught myself duplicating even more with the #229 feature.

The purpose is create a internal package to be used by the cli to manage it and reduce code duplication.

### Checklist for proposal reviewer

- [ ] The proposal is clear on what changes need to be made in the code base.

- [ ] The proposal is clear on the architectural changes that will be affected.

- [ ] The proposal is clear on the breaking changes and how to mitigate them.

- [ ] The proposal is clear on what the deployment/configuration strategy is.

- [ ] The proposal is clean on how testing will validate the feature implementation.

### Reviewers

- [ ] Reviewed by a team member of the @calyptia/backend team
- [ ] Reviewed by a product owner.

### Disclaimers

- If there are multiple solutions, please present each one separately.
- Save comparisons for the very end.
- Please do not create a Pull Request without creating an issue first.
- Any change needs to be discussed before proceeding.
- Failure to do so may result in the rejection of the pull request.

### PR DESCRIPTION
**Pull request by gabrielbussolo**
Signed-off-by: Gabriel Bussolo <gabriel@calyptia.com>

# Summary of this proposal

This PR intends to resolve #299 

Creates a package to be used to save small data.

The usability basically is `localData.Save(key, data)`, `localData.Get(key)`, `localData.Delete(key)`, it will try to save on the os keyring if its not available it will save on a local file like we were doing before, the refactor was around the usability and the code duplication, not around the functionality itself so it remains the same, just a bit **more reusable**, **documented** and with **tests**.

If it looks good for everyone it will be used on the #256 implementation also

ps: the package is `internal` to prevent imports on other projects and with it we dont need to provide support, but i put it on the `platform` package what means that one day it can be decoupled from the cli to be used by other projects. (according with package oriented design)

## Issue(s) number(s)

*Disclaimer: Please do not create a Pull Request without creating an issue first.*

## Checklist for submitter

- [x] My PR has a related issue/bug number.
- [ ] My PR provides tests
  - [ ] Integration tests are added/passes
  - [x] Unit tests are added/passes
  - [ ] End-to-end tests added/passes
  - [ ] Static code check added/passes
  - [ ] Linting on documentation added/passes
  - [ ] Does not affect code coverage stats
- [ ] My PR requires updating dependencies
- [ ] My PR has the documentation changes required.

## Checklist for reviewer

- [ ] The proposal fixes a bug/issue or implements a new feature that is well described.
- [ ] The proposal has sufficient test cases that covers the changes.
  - [ ] If changes an API, it does not break backwards compatibility (-1 major version).
- [ ] If integration is required, the proposal has integration tests.
- [ ] The proposal does not break coverage stats
- [ ] The proposal has the required documentation changes.

## Backport

- [ ] Backport to the latest stable release.

### CODE REVIEW

**Comment by patrick-stephens** [comment]
Is there no existing standard package to do this @gabrielbussolo ?

**Comment by gabrielbussolo** [comment]
> Is there no existing standard package to do this @gabrielbussolo ?

either on golang standard library or our packages we dont have a solution like we have been doing here, we have been repeating this code on cli codebase so i basically "generify"  it and externalized to package to be more reusable

**Comment by patrick-stephens** [comment]
For me it was more is there a standard or good existing package people generally use rather than make our own but I'm not averse to it.

**Comment by nicolasparada** [comment]
@patrick-stephens https://github.com/99designs/keyring this one is better, securer and has a fallback mechanism to store as plain file. But would break existing flow, that's why we haven't switched to it.

File: cmd/calyptia/config_url.go [hunk 1]
```diff
@@ -3,17 +3,14 @@ package main
...
-	"path/filepath"
-
 	"github.com/spf13/cobra"
-	"github.com/zalando/go-keyring"
+	"net/url"
 )
 
 var errURLNotFound = errors.New("url not found")
 
+const UrlToken = "base_url"
```
Comment (nicolasparada) [inline cmd/calyptia/config_url.go:12]:
```suggestion
const keyBaseURL = "base_url"
```

File: cmd/calyptia/config_token.go [hunk 2]
```diff
@@ -6,6 +6,8 @@ import (
 	"github.com/spf13/cobra"
 )
 
+const TokenKey = "token"
```
Comment (nicolasparada) [inline cmd/calyptia/config_token.go:9]:
```suggestion
const keyToken = "token"
```

**Review by nicolasparada — APPROVED** [review]
Please rename some of the keys. Prefix all of them with `Key`, ex `KeyFoo`.

Also sort imports and stuff.

```bash
goimports -w -local github.com/calyptia/ .
```

**Review by nicolasparada — CHANGES_REQUESTED** [review]
Sorry, I meant to request changes.

**Review by nicolasparada — CHANGES_REQUESTED** [review]
Oh, didn't notice the package name.
There is no need for all that subpackaging.

You can create a folder called `keyring` and but the code there instead of `internal/platform/local-data`.

The new `keyring` package is basically a wrapper over keyring but will local filesystem fallback.

**Comment by gabrielbussolo** [comment]
> Oh, didn't notice the package name. There is no need for all that subpackaging.
> 
> You can create a folder called `keyring` and but the code there instead of `internal/platform/local-data`.
> 
> The new `keyring` package is basically a wrapper over keyring but will local filesystem fallback.

the reason for that is on the pr description 
> ps: the package is internal to prevent imports on other projects and with it we don't need to provide support, but i put it on > the platform package what means that one day it can be decoupled from the cli to be used by other projects. (according > > with package oriented design)

but to resume basically isn't from the scope of calyptia/cli keep a keyring api for external use, lets keep it on internal.

**Comment by nicolasparada** [comment]
That is true for almost all packages but we are not using `internal` in any of them.
I wouldn't use unless there is a strong reason, which we don't have.

And platform is just meaningless, it's not idiomatic at all.

`local-data` is also not a way to name a package due to the dash. It can be `localdata` if you don't like simply `keyring`.

File: cmd/calyptia/auth.go [hunk 3]
```diff
@@ -6,100 +6,12 @@ import (
...
 	"fmt"
-	"io/fs"
-	"os"
-	"path/filepath"
-
-	"github.com/zalando/go-keyring"
 )
 
 const serviceName = "cloud.calyptia.com"
+const backUpFolder = ".calyptia"
```
Comment (patrick-stephens) [inline cmd/calyptia/auth.go:12]: Things like this need to be captured in the docs - we need to ensure complete removal for package install for example. Where exactly is this folder (under a home area for a user or in the calyptia area)?

Please if you're adding extra directories or files then make sure you raise an issue on calyptia/core-images so we track the file/directory properly for packages. It's best practice and quite often a requirement for more secure systems.

**Review by nicolasparada — CHANGES_REQUESTED** [review]
There is no need for a `pkg` directory, it's verbose to have a package called "package". Does not provide anything.

File: cmd/calyptia/auth.go (hunk 3 above)
Comment (patrick-stephens) [reply cmd/calyptia/auth.go:12]: What happens if we run this as a user with no home directory?
Can we make it configurable?

**Review by nicolasparada — APPROVED** [review]

**Pull request merged**